adapters/zemanta/zemanta.go

```diff
@@ -0,0 +1,147 @@
+package zemanta
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+// Builder builds a new instance of the Zemanta adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	reqCopy := *request
+
+	var errs []error
+	var zemantaExt openrtb_ext.ExtImpZemanta
+	for i := 0; i < len(reqCopy.Imp); i++ {
+		imp := reqCopy.Imp[i]
+
+		var bidderExt adapters.ExtImpBidder
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if err := json.Unmarshal(bidderExt.Bidder, &zemantaExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		imp.TagID = zemantaExt.TagId
+		reqCopy.Imp[i] = imp
```

Comment:
It looks to me that you don't want to modify the original request, since you're making modifications on a shallow `reqCopy` copy. Not a problem.

I'd like to point out this code is modifying the original request at this line. The `reqCopy.Imp` is a Go Slice which uses the same underlying memory as the original `request.Imp`. The `Imp` copy is unique to each adapter, so there's no issue. I just wanted you to be aware. If you'd like to leave the original request alone, you'll want to clone the `reqCopy.Imp` slice before the loop.

Reply:
Isn't it prohibited to change the original request? 
The documentation in https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html states:
> If your adapter needs to alter any data referenced by a pointer then you must first make a shallow copy. The only exception is for request.Imp and its elements, as these are already shallow copies.

Reply:
The request provided to your adapter is already a shallow copy. You may change anything at the root level of the request or anything at the root level of an imp. If you want to change a variable within a pointer, then that's shared memory and you need to make a copy first.. as you're doing correctly with site and app.

The consequence is the request given to MakeRequests is also given to MakeBids, so if you modify the request in MakeRequests then MakeBids will work off the modified request, and that is often times ok.

I'll update the doc to explain that the request is also a shallow copy. Thank you for pointing out that inconsistency.